src/dat.rs

```diff
@@ -0,0 +1,208 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+//
+// Copyright (c) DUSK NETWORK. All rights reserved.
+
+use blake3::Hash;
+use dusk_bytes::DeserializableSlice;
+
+use crate::crypto::decrypt;
+use crate::store;
+use crate::Error;
+
+/// Binary prefix for old Dusk wallet files
+pub const OLD_MAGIC: u32 = 0x1d0c15;
+/// Binary prefix for new binary file format
+pub const MAGIC: u32 = 0x72736b;
+/// Specifies the encoding used to save files
+pub const VERSION: &[u8] = &[2, 0];
+
+/// (Major, Minor, Patch, Pre, Pre-Higher)
+type Version = (u8, u8, u8, u8, bool);
+
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub(crate) enum DatFileVersion {
+    /// Legacy the oldest format
+    Legacy,
+    /// Preciding legacy, we have the old one
+    OldWalletCli(Version),
+    /// The newest one. All new saves are saved in this file format
+    RuskBinaryFileFormat(Version),
+}
+
+pub(crate) fn get_seed_and_address(
+    file: DatFileVersion,
+    mut bytes: Vec<u8>,
+    pwd: Hash,
+) -> Result<(store::Seed, u8), Error> {
+    match file {
+        DatFileVersion::Legacy => {
+            if bytes[1] == 0 && bytes[2] == 0 {
+                bytes.drain(..3);
+            }
+
+            bytes = decrypt(&bytes, pwd)?;
+
+            // get our seed
+            let seed = store::Seed::from_reader(&mut &bytes[..])
+                .map_err(|_| Error::WalletFileCorrupted)?;
+
+            Ok((seed, 1))
+        }
+        DatFileVersion::OldWalletCli((major, minor, _, _, _)) => {
+            bytes.drain(..5);
+
+            let result: Result<(store::Seed, u8), Error> = match (major, minor)
+            {
+                (1, 0) => {
+                    let content = decrypt(&bytes, pwd)?;
+                    let mut buff = &content[..];
+
+                    let seed = store::Seed::from_reader(&mut buff)
+                        .map_err(|_| Error::WalletFileCorrupted)?;
+
+                    Ok((seed, 1))
+                }
+                (2, 0) => {
+                    let content = decrypt(&bytes, pwd)?;
+                    let mut buff = &content[..];
+
+                    // extract seed
+                    let seed = store::Seed::from_reader(&mut buff)
+                        .map_err(|_| Error::WalletFileCorrupted)?;
+
+                    // extract addresses count
+                    Ok((seed, buff[0]))
+                }
+                _ => Err(Error::UnknownFileVersion(major, minor)),
+            };
+
+            result
+        }
+        DatFileVersion::RuskBinaryFileFormat(_) => {
+            let rest = bytes.get(12..(12 + 96));
+            if let Some(rest) = rest {
+                let content = decrypt(rest, pwd)?;
```

Comment:
The new RuskBinaryFormat is not encrypted with a Blake hash, so you cannot pass a `blake3::Hash` struct.

Consider changing the inner `decrypt` method to accept `&[u8]` (that is exactly the type needed by `Aes256Cbc::new_from_slices`) and use it accordingly to the matching DatFileVersion

Reply:
Fixed with #171
@herr-seppia